src/neovim.rs

```diff
@@ -108,13 +109,98 @@ where
     let (sender, receiver) = unbounded();
     let fut = future::try_join(
       req.clone().io_loop(reader, sender),
-      req.clone().handler_loop(handler, receiver)
+      req.clone().handler_loop(handler, receiver),
     )
     .map_ok(|_| ());
 
     (req, fut)
   }
 
+  /// Create a new instance, immediately send a handshake message and
+  /// wait for the response. Unlike `new`, this function is tolerant to extra
+  /// data in the reader before the handshake response is received.
+  pub async fn handshake<H, R>(
+    mut reader: R,
+    writer: W,
+    handler: H,
+  ) -> Result<
+    (
+      Neovim<<H as Handler>::Writer>,
+      impl Future<Output = Result<(), Box<LoopError>>>,
+    ),
+    Box<HandshakeError>,
+  >
+  where
+    R: AsyncRead + Send + Unpin + 'static,
+    H: Handler<Writer = W> + Spawner,
+  {
+    let instance = Neovim {
+      writer: Arc::new(Mutex::new(writer)),
+      msgid_counter: Arc::new(AtomicU64::new(0)),
+      queue: Arc::new(Mutex::new(Vec::new())),
+    };
+
+    let msgid = instance.msgid_counter.fetch_add(1, Ordering::SeqCst);
+    let magic = "nvim-rsMagicHandshakeMarkernvim-rs";
+    // Nvim encodes fixed size strings with a length of 20-31 bytes wrong, so
+    // avoid that
+    assert!(magic.len() < 20 || magic.len() > 31);
+
+    let req = RpcMessage::RpcRequest {
+      msgid,
+      method: "nvim_exec_lua".to_owned(),
+      params: call_args![format!("return '{magic}'"), Vec::<Value>::new()],
+    };
+    model::encode(instance.writer.clone(), req).await?;
+
+    let expected_resp = RpcMessage::RpcResponse {
+      msgid,
+      error: rmpv::Value::Nil,
+      result: rmpv::Value::String(magic.into()),
+    };
+    let mut expected_data = Vec::new();
+    model::encode_sync(&mut expected_data, expected_resp)
+      .expect("Encoding static data can't fail");
+    let mut actual_data = Vec::new();
+    let mut start = 0;
+    let mut end = 0;
+    while end - start != expected_data.len() {
+      actual_data.resize(start + expected_data.len(), 0);
+
+      let bytes_read =
+        reader
+          .read(&mut actual_data[start..])
+          .await
+          .map_err(|err| {
+            (
+              err,
+              String::from_utf8_lossy(&actual_data[..end]).to_string(),
+            )
+          })?;
+      if bytes_read == 0 {
```

Comment:
Please add a comment clarifying that tokio's async reader only returns `Ok(0)` in case the stream was closed (I find this important because that's quite different to sync readers).